### PR TITLE
refactor!: FunctionBuilder takes impl Into<PolyFuncType>

### PR DESCRIPTION
**Pull request by acl-cqc**
Currently FunctionBuilder::new and the `define_function` method both take a PolyFuncType; this PR changes that to a more-general `impl Into<PolyFuncType>`.

BREAKING CHANGE: clients calling these methods with a FunctionType's `.into()` will fail to compile with a "type annotations required" error, which can be fixed by deleting the `.into()`.

### CODE REVIEW

File: hugr-core/src/builder.rs [hunk 1]
```diff
@@ -59,7 +59,7 @@
 //!     let _circuit_handle = {
 //!         let mut dfg = module_builder.define_function(
 //!             "circuit",
-//!             FunctionType::new(vec![BOOL_T, BOOL_T], vec![BOOL_T, BOOL_T]).into(),
+//!             FunctionType::new_endo(vec![BOOL_T, BOOL_T]),
```
Comment (acl-cqc) [inline hugr-core/src/builder.rs:62]: `new_endo` here is a driveby but seemed a reasonable thing to demonstrate in a doctest

**Review by cqc-alec — APPROVED** [review]
Nice.
